web/packages/teleport/src/UnifiedResources/ResourceCard.tsx

```diff
@@ -54,60 +52,146 @@ const ResTypeIconBox = styled(Box)`
+          // would move down along with expanding the labels. So we apply a
+          // carefully measured top margin instead.
+          mt="16px"
+        />
+        {/* MinWidth is important to prevent descriptions from overflowing. */}
+        <Flex flexDirection="column" flex="1" minWidth="0" ml={3} gap={1}>
+          <Flex flexDirection="row" alignItems="start">
+            <SingleLineBox flex="1" title={name}>
+              <Text typography="h5">{name}</Text>
             </SingleLineBox>
-          )}
-          {description.secondary && (
-            <SingleLineBox ml={2} title={description.secondary}>
-              <Text typography="body2" color="text.muted">
-                {description.secondary}
-              </Text>
-            </SingleLineBox>
-          )}
-        </Flex>
-        <Flex gap={1}>
-          {resource.labels.map(({ name, value }) => {
-            const label = `${name}: ${value}`;
-            return (
-              <TruncatingLabel key={label} title={label} kind="secondary">
-                {label}
-              </TruncatingLabel>
-            );
-          })}
+            <ButtonBorder size="small">Connect</ButtonBorder>
+          </Flex>
+          <Flex flexDirection="row" alignItems="center">
+            <ResTypeIconBox>
+              <ResTypeIcon size={18} />
+            </ResTypeIconBox>
+            {description.primary && (
+              <SingleLineBox ml={1} title={description.primary}>
+                <Text typography="body2" color="text.slightlyMuted">
+                  {description.primary}
+                </Text>
+              </SingleLineBox>
+            )}
+            {description.secondary && (
+              <SingleLineBox ml={2} title={description.secondary}>
+                <Text typography="body2" color="text.muted">
+                  {description.secondary}
+                </Text>
+              </SingleLineBox>
+            )}
+          </Flex>
+          <LabelsContainer showAll={showAllLabels}>
+            <LabelsInnerContainer ref={labelsInnerContainer}>
+              <MoreLabelsButton
+                style={{
+                  visibility:
+                    showMoreLabelsButton && !showAllLabels
+                      ? 'visible'
+                      : 'hidden',
+                }}
+                onClick={onMoreLabelsClick}
+              >
+                + {numMoreLabels} more
+              </MoreLabelsButton>
+              {resource.labels.map(label => {
+                const { name, value } = label;
+                const labelText = `${name}: ${value}`;
+                return (
+                  <ResourceLabel
+                    key={labelText}
+                    title={labelText}
+                    onClick={() => onLabelClick?.(label)}
+                    kind="secondary"
+                    data-is-label=""
+                  >
+                    {labelText}
+                  </ResourceLabel>
+                );
```

Comment:
We've ran into issues before with duplicate labels causing some weird bugs that would infinitely spawn more and more labels as the user used the UI. [check it out](https://github.com/gravitational/teleport.e/pull/722).
Generally speaking, duplicate labels _shouldn't_ happen but they can as we've seen so we need to add an extra identifier.

Reply:
Hmm, the change that you mentioned is about duplicate _resources_, not labels, isn't it? The labels are just controls that are used for display.

Also, that PR contains a very weird corner case: if I had resources called `foo1` on the 5th position in the list, and a resource called `foo` on the 15th position, they'd share the same key. :)

In the hindsight, I should probably also use something like `JSON.stringify([name, value])` for dodging all possible corner cases, since I still may run into a bug if someone puts a colon in the label name or value.